doc/adrs/011_refactor_incremental_layout.adoc

```diff
@@ -0,0 +1,231 @@
+ELK manage this well, with long-time studied theories in application. That's how it come with solutions with no overlap, minimal crosses between edges and so on.
+
+The purpose of the incremental layout is quite different as another factor comes at play: ergonomy.
+We want the user not to be disturbed by the changes, i.e. the new diagram must look like the previous one. The only changes shown must be strongly related to the performed actions.
+The consequence of that is for instance when computing edges: the "keep crossings minimal" injonctions is no longer relevant, what we want here is to make the minimal changes to the bendpoints list to make the edge connected with the new positions, and still respect the edge layout mode.
+
+=== Layout data structure
+
+As for the ELK layout, we first need to create a mutable structure of the diagram, which will hold the layout values and ease the navigation.
+A conversion of the existing diagram will produce a data structure similar to the diagram, initialized with existing values.
+This structure will then be modified by an incremental layout engine which includes inputs from the frontend (moves, resizes, creations) and fix any missing layouts / layouts issues (overlaps, wrong sizes).
+Finally this structure will be used to produce a new immutable diagram with the correct layout.
+
+=== Layout computation
+
+The incremental layout engine will implement the following process:
+
+```
+layout(diagram) {
+	diagram.children.forEach(layoutNode(node))
+
+	resolveOverlaps(diagram)
+	resolveContainmentIssues(diagram)
+
+	diagram.edges.forEach(
+		if (hasChanged(edge.source) || hasChanged(edge.target)) {
+            		layoutEdge(edge)
+		}
+	)
+}
+
+layoutNode(node) {
+	diagram.border.forEach(layoutNode(borderNode))
+	diagram.children.forEach(layoutNode(childNode))
+
+	if (hasBeenResized(node)) {
+		node.setSize(newSize)
+	} else if (isNewNode(node)) {
+		node.setSize(computeNewSize)
+	}
+
+	if (hasBeenMoved(node)) {
+		node.setPosition(newPosition)
+	} else if (isNewNode(node)) {
+		node.setPosition(computeNewPosition)
+	}
+
+        resolveOverlaps(node)
+	resolveContainmentIssues(node)
+
+	node.label.setPosition(computeNewLabelPosition)
+    }
+
+layoutEdge(edge) {
+        edge.setRoutingPoints(computeRoutingPoints)
+	computeEdgeLabels(edge)
+}
+```
+
+The underlying sections describes the layout computation steps for each kind of graphical component.
+For each computation, we show the dependencies of it, i.e. what factor leads to a recomputation of the previous values.
+
+==== Node sizes and positions
+
+===== Dependencies
+
+A element or any of its children/siblings has been moved/resized.
+
+===== Computation
+
+The computation of the nodes size and positions is the key point of the algorithm : any change made by the algorithm itself, to follow a move/resize from the frontend, can imply another indirect move/resize the can imply other changes.
```

Comment:
that can imply other changes.

Reply:
Fixed in https://github.com/wpiers/sirius-components/commit/62f8cf80e053d114a0f7655bd26db3318e521999